# Importing Data Using Pandas - Lab

## Introduction

In this lab, you'll get some practice with loading files with summary or metadata, and if you find that easy, the optional "level up" content covers loading data from a corrupted csv file!

## Objectives
You will be able to:
* Import data from csv files and Excel files
* Understand and explain key arguments for imports
* Save information to csv and Excel files
* Access data within a Pandas DataFrame (print() and .head())

##  Loading Files with Summary or Meta Data

Load either of the files Zipcode_Demos.csv or Zipcode_Demos.xlsx. What's going on with this dataset? Clean it up into a useable format and describe the nuances of how the data is currently formatted.

All data files are stored in a folder titled 'Data'.

### Loading a csv file 

In [51]:
import pandas as pd
df = pd.read_csv('data\\Zipcode_Demos.csv')
df.iloc[46]

0                                                      47
Average Statistics                      JURISDICTION NAME
Unnamed: 2                             COUNT PARTICIPANTS
Unnamed: 3                                   COUNT FEMALE
Unnamed: 4                                 PERCENT FEMALE
Unnamed: 5                                     COUNT MALE
Unnamed: 6                                   PERCENT MALE
Unnamed: 7                           COUNT GENDER UNKNOWN
Unnamed: 8                         PERCENT GENDER UNKNOWN
Unnamed: 9                             COUNT GENDER TOTAL
Unnamed: 10                          PERCENT GENDER TOTAL
Unnamed: 11                        COUNT PACIFIC ISLANDER
Unnamed: 12                      PERCENT PACIFIC ISLANDER
Unnamed: 13                         COUNT HISPANIC LATINO
Unnamed: 14                       PERCENT HISPANIC LATINO
Unnamed: 15                         COUNT AMERICAN INDIAN
Unnamed: 16                       PERCENT AMERICAN INDIAN
Unnamed: 17   

In [38]:
prev_count = 10**3
for row in df.index:
    count = 0
    for entry in df.iloc[row].isnull():#df.iloc[row].isnull() is a list of boolean values
        if entry:
            count += 1
    if count != prev_count and row!=0:#checking if the number or null values in a row changes at some specific one other than 1st row
        print('On row {} there are {} null values. The previous row had {} null values.'.format(row, count, prev_count))
    prev_count = count

On row 1 there are 44 null values. The previous row had 45 null values.
On row 46 there are 0 null values. The previous row had 44 null values.


In [41]:
 df[0:47]

0                   Average Statistics          Unnamed: 2    Unnamed: 3  \
0    1                                  NaN                   0           NaN   
1    2                    JURISDICTION NAME             10005.8           NaN   
2    3                   COUNT PARTICIPANTS                 9.4           NaN   
3    4                         COUNT FEMALE                 4.8           NaN   
4    5                       PERCENT FEMALE               0.404           NaN   
5    6                           COUNT MALE                 4.6           NaN   
6    7                         PERCENT MALE               0.396           NaN   
7    8                 COUNT GENDER UNKNOWN                   0           NaN   
8    9               PERCENT GENDER UNKNOWN                   0           NaN   
9   10                   COUNT GENDER TOTAL                 9.4           NaN   
10  11                 PERCENT GENDER TOTAL                  80           NaN   
11  12               COUNT PACIFIC ISLANDER                   0           NaN   
12  13             PERCENT PACIFIC ISLANDER                   0           NaN   
13  14                COUNT HISPANIC LATINO                   2           NaN   
14  15              PERCENT HISPANIC LATINO               0.105           NaN   
15  16                COUNT AMERICAN INDIAN                   0           NaN   
16  17              PERCENT AMERICAN INDIAN                   0           NaN   
17  18             COUNT ASIAN NON HISPANIC                 3.6           NaN   
18  19           PERCENT ASIAN NON HISPANIC               0.437           NaN   
19  20             COUNT WHITE NON HISPANIC                 0.8           NaN   
20  21           PERCENT WHITE NON HISPANIC               0.036           NaN   
21  22             COUNT BLACK NON HISPANIC                 2.6           NaN   
22  23           PERCENT BLACK NON HISPANIC               0.181           NaN   
23  24                COUNT OTHER ETHNICITY                 0.4           NaN   
24  25              PERCENT OTHER ETHNICITY                0.04           NaN   
25  26              COUNT ETHNICITY UNKNOWN                   0           NaN   
26  27            PERCENT ETHNICITY UNKNOWN                   0           NaN   
27  28                COUNT ETHNICITY TOTAL                 9.4           NaN   
28  29              PERCENT ETHNICITY TOTAL                79.9           NaN   
29  30       COUNT PERMANENT RESIDENT ALIEN                 0.5           NaN   
30  31     PERCENT PERMANENT RESIDENT ALIEN               0.061           NaN   
31  32                     COUNT US CITIZEN                 8.9           NaN   
32  33                   PERCENT US CITIZEN               0.739           NaN   
33  34           COUNT OTHER CITIZEN STATUS                   0           NaN   
34  35         PERCENT OTHER CITIZEN STATUS                   0           NaN   
35  36         COUNT CITIZEN STATUS UNKNOWN                   0           NaN   
36  37       PERCENT CITIZEN STATUS UNKNOWN                   0           NaN   
37  38           COUNT CITIZEN STATUS TOTAL                 9.4           NaN   
38  39         PERCENT CITIZEN STATUS TOTAL                  80           NaN   
39  40     COUNT RECEIVES PUBLIC ASSISTANCE                 2.3           NaN   
40  41   PERCENT RECEIVES PUBLIC ASSISTANCE               0.151           NaN   
41  42    COUNT NRECEIVES PUBLIC ASSISTANCE                 7.1           NaN   
42  43  PERCENT NRECEIVES PUBLIC ASSISTANCE               0.649           NaN   
43  44      COUNT PUBLIC ASSISTANCE UNKNOWN                   0           NaN   
44  45    PERCENT PUBLIC ASSISTANCE UNKNOWN                   0           NaN   
45  46        COUNT PUBLIC ASSISTANCE TOTAL                 9.4           NaN   
46  47                    JURISDICTION NAME  COUNT PARTICIPANTS  COUNT FEMALE   

        Unnamed: 4  Unnamed: 5    Unnamed: 6            Unnamed: 7  \
0              NaN         NaN           NaN  

In [45]:
df1 = pd.read_csv('data\\Zipcode_Demos.csv', skiprows = [1], nrows = 45, usecols = [0,1,2])
df1.tail()

0                   Average Statistics  Unnamed: 2
40  42    COUNT NRECEIVES PUBLIC ASSISTANCE       7.100
41  43  PERCENT NRECEIVES PUBLIC ASSISTANCE       0.649
42  44      COUNT PUBLIC ASSISTANCE UNKNOWN       0.000
43  45    PERCENT PUBLIC ASSISTANCE UNKNOWN       0.000
44  46        COUNT PUBLIC ASSISTANCE TOTAL       9.400

In [49]:
df2 = pd.read_csv('data\\Zipcode_Demos.csv', skiprows = 47)#Then the last skipped row (47th one, index = 46) becomes header. 
df2

47  JURISDICTION NAME  COUNT PARTICIPANTS  COUNT FEMALE  PERCENT FEMALE  \
0  48              10001                  44            22            0.50   
1  49              10002                  35            19            0.54   
2  50              10003                   1             1            1.00   
3  51              10004                   0             0            0.00   
4  52              10005                   2             2            1.00   
5  53              10006                   6             2            0.33   
6  54              10007                   1             0            0.00   
7  55              10009                   2             0            0.00   
8  56              10010                   0             0            0.00   
9  57              10011                   3             2            0.67   

   COUNT MALE  PERCENT MALE  COUNT GENDER UNKNOWN  PERCENT GENDER UNKNOWN  \
0          22          0.50                     0                       0   
1          16          0.46                     0                       0   
2           0          0.00                     0                       0   
3           0          0.00                     0                       0   
4           0          0.00                     0                       0   
5           4          0.67                     0                       0   
6           1          1.00                     0                       0   
7           2          1.00                     0                       0   
8           0          0.00                     0                       0   
9           1          0.33                     0                       0   

   COUNT GENDER TOTAL               ...                 \
0                  44               ...                  
1                  35               ...                  
2                   1               ...                  
3                   0               ...                  
4                   2               ...                  
5                   6               ...                  
6                   1               ...                  
7                   2               ...                  
8                   0               ...                  
9                   3               ...                  

   COUNT CITIZEN STATUS TOTAL  PERCENT CITIZEN STATUS TOTAL  \
0                          44                           100   
1                          35                           100   
2                           1                           100   
3                           0                             0   
4                           2                           100   
5                           6                           100   
6                           1                           100   
7                           2                           100   
8                           0                             0   
9                           3                           100   

   COUNT RECEIVES PUBLIC ASSISTANCE  PERCENT RECEIVES PUBLIC ASSISTANCE  \
0                                20                                0.45   
1                                 2                                0.06   
2                                 0                                0.00   
3                                 0                                0.00   
4                                 0                                0.00   
5                                 0                                0.00   
6                                 1                                1.00   
7                                 0                                0.00   
8                                 0                                0.00   
9                                 0                                0.00   

   COUNT NRECEIVES PUBLIC ASSISTANCE  PERCENT NRECEIVES PUBLIC ASSISTANCE  \
0                                 24                        

### Loading an excel file 

In [8]:
workbook = pd.ExcelFile('data\\Zipcode_Demos.xlsx')
workbook.sheet_names


['Sheet1']

## Note that excel file and csv file were loaded differently 

In [53]:
df3 = pd.read_excel('data\\Zipcode_Demos.xlsx')
df3

0          Unnamed: 1    Unnamed: 2  \
0                    Average Statistics                 NaN           NaN   
1                                   NaN                   0           NaN   
2                     JURISDICTION NAME             10005.8           NaN   
3                    COUNT PARTICIPANTS                 9.4           NaN   
4                          COUNT FEMALE                 4.8           NaN   
5                        PERCENT FEMALE               0.404           NaN   
6                            COUNT MALE                 4.6           NaN   
7                          PERCENT MALE               0.396           NaN   
8                  COUNT GENDER UNKNOWN                   0           NaN   
9                PERCENT GENDER UNKNOWN                   0           NaN   
10                   COUNT GENDER TOTAL                 9.4           NaN   
11                 PERCENT GENDER TOTAL                  80           NaN   
12               COUNT PACIFIC ISLANDER                   0           NaN   
13             PERCENT PACIFIC ISLANDER                   0           NaN   
14                COUNT HISPANIC LATINO                   2           NaN   
15              PERCENT HISPANIC LATINO               0.105           NaN   
16                COUNT AMERICAN INDIAN                   0           NaN   
17              PERCENT AMERICAN INDIAN                   0           NaN   
18             COUNT ASIAN NON HISPANIC                 3.6           NaN   
19           PERCENT ASIAN NON HISPANIC               0.437           NaN   
20             COUNT WHITE NON HISPANIC                 0.8           NaN   
21           PERCENT WHITE NON HISPANIC               0.036           NaN   
22             COUNT BLACK NON HISPANIC                 2.6           NaN   
23           PERCENT BLACK NON HISPANIC               0.181           NaN   
24                COUNT OTHER ETHNICITY                 0.4           NaN   
25              PERCENT OTHER ETHNICITY                0.04           NaN   
26              COUNT ETHNICITY UNKNOWN                   0           NaN   
27            PERCENT ETHNICITY UNKNOWN                   0           NaN   
28                COUNT ETHNICITY TOTAL                 9.4           NaN   
29              PERCENT ETHNICITY TOTAL                79.9           NaN   
30       COUNT PERMANENT RESIDENT ALIEN                 0.5           NaN   
31     PERCENT PERMANENT RESIDENT ALIEN               0.061           NaN   
32                     COUNT US CITIZEN                 8.9           NaN   
33                   PERCENT US CITIZEN               0.739           NaN   
34           COUNT OTHER CITIZEN STATUS                   0           NaN   
35         PERCENT OTHER CITIZEN STATUS                   0           NaN   
36         COUNT CITIZEN STATUS UNKNOWN                   0           NaN   
37       PERCENT CITIZEN STATUS UNKNOWN                   0           NaN   
38           COUNT CITIZEN STATUS TOTAL                 9.4           NaN   
39         PERCENT CITIZEN STATUS TOTAL                  80           NaN   
40     COUNT RECEIVES PUBLIC ASSISTANCE                 2.3           NaN   
41   PERCENT RECEIVES PUBLIC ASSISTANCE               0.151           NaN   
42    COUNT NRECEIVES PUBLIC ASSISTANCE                 7.1           NaN   
43  PERCENT NRECEIVES PUBLIC ASSISTANCE               0.649           NaN   
44      COUNT PUBLIC ASSISTANCE UNKNOWN                   0           NaN   
45    PERCENT PUBLIC ASSISTANCE UNKNOWN                   0           NaN   
46        COUNT PUBLIC ASSISTANCE TOTAL                 9.4           NaN   
47                    JURISDICTION NAME  COUNT PARTICIPANTS  COUNT FEMALE   
48                                10001                  44            22   
49                                10002                  35            19   
50                                10003                   1             1   
51                                

In [60]:
 df4 = pd.read_excel('data\\Zipcode_Demos.xlsx', sheet_name = 0, skiprows =[0,2], nrows =45 ,usecols = [0,1])
#df4 = pd.read_excel('data\\Zipcode_Demos.xlsx', sheet_name = 0, skiprows = 1)
df4

Average Statistics          Unnamed: 1
0                     JURISDICTION NAME             10005.8
1                    COUNT PARTICIPANTS                 9.4
2                          COUNT FEMALE                 4.8
3                        PERCENT FEMALE               0.404
4                            COUNT MALE                 4.6
5                          PERCENT MALE               0.396
6                  COUNT GENDER UNKNOWN                   0
7                PERCENT GENDER UNKNOWN                   0
8                    COUNT GENDER TOTAL                 9.4
9                  PERCENT GENDER TOTAL                  80
10               COUNT PACIFIC ISLANDER                   0
11             PERCENT PACIFIC ISLANDER                   0
12                COUNT HISPANIC LATINO                   2
13              PERCENT HISPANIC LATINO               0.105
14                COUNT AMERICAN INDIAN                   0
15              PERCENT AMERICAN INDIAN                   0
16             COUNT ASIAN NON HISPANIC                 3.6
17           PERCENT ASIAN NON HISPANIC               0.437
18             COUNT WHITE NON HISPANIC                 0.8
19           PERCENT WHITE NON HISPANIC               0.036
20             COUNT BLACK NON HISPANIC                 2.6
21           PERCENT BLACK NON HISPANIC               0.181
22                COUNT OTHER ETHNICITY                 0.4
23              PERCENT OTHER ETHNICITY                0.04
24              COUNT ETHNICITY UNKNOWN                   0
25            PERCENT ETHNICITY UNKNOWN                   0
26                COUNT ETHNICITY TOTAL                 9.4
27              PERCENT ETHNICITY TOTAL                79.9
28       COUNT PERMANENT RESIDENT ALIEN                 0.5
29     PERCENT PERMANENT RESIDENT ALIEN               0.061
30                     COUNT US CITIZEN                 8.9
31                   PERCENT US CITIZEN               0.739
32           COUNT OTHER CITIZEN STATUS                   0
33         PERCENT OTHER CITIZEN STATUS                   0
34         COUNT CITIZEN STATUS UNKNOWN                   0
35       PERCENT CITIZEN STATUS UNKNOWN                   0
36           COUNT CITIZEN STATUS TOTAL                 9.4
37         PERCENT CITIZEN STATUS TOTAL                  80
38     COUNT RECEIVES PUBLIC ASSISTANCE                 2.3
39   PERCENT RECEIVES PUBLIC ASSISTANCE               0.151
40    COUNT NRECEIVES PUBLIC ASSISTANCE                 7.1
41  PERCENT NRECEIVES PUBLIC ASSISTANCE               0.649
42      COUNT PUBLIC ASSISTANCE UNKNOWN                   0
43    PERCENT PUBLIC ASSISTANCE UNKNOWN                   0
44        COUNT PUBLIC ASSISTANCE TOTAL                 9.4
45                    JURISDICTION NAME  COUNT PARTICIPANTS

In [64]:
df5 = pd.read_excel('data\\Zipcode_Demos.xlsx', sheet_name = 0, skiprows =48)
df5

JURISDICTION NAME  COUNT PARTICIPANTS  COUNT FEMALE  PERCENT FEMALE  \
0              10001                  44            22            0.50   
1              10002                  35            19            0.54   
2              10003                   1             1            1.00   
3              10004                   0             0            0.00   
4              10005                   2             2            1.00   
5              10006                   6             2            0.33   
6              10007                   1             0            0.00   
7              10009                   2             0            0.00   
8              10010                   0             0            0.00   
9              10011                   3             2            0.67   

   COUNT MALE  PERCENT MALE  COUNT GENDER UNKNOWN  PERCENT GENDER UNKNOWN  \
0          22          0.50                     0                       0   
1          16          0.46                     0                       0   
2           0          0.00                     0                       0   
3           0          0.00                     0                       0   
4           0          0.00                     0                       0   
5           4          0.67                     0                       0   
6           1          1.00                     0                       0   
7           2          1.00                     0                       0   
8           0          0.00                     0                       0   
9           1          0.33                     0                       0   

   COUNT GENDER TOTAL  PERCENT GENDER TOTAL               ...                 \
0                  44                   100               ...                  
1                  35                   100               ...                  
2                   1                   100               ...                  
3                   0                     0               ...                  
4                   2                   100               ...                  
5                   6                   100               ...                  
6                   1                   100               ...                  
7                   2                   100               ...                  
8                   0                     0               ...                  
9                   3                   100               ...                  

   COUNT CITIZEN STATUS TOTAL  PERCENT CITIZEN STATUS TOTAL  \
0                          44                           100   
1                          35                           100   
2                           1                           100   
3                           0                             0   
4                           2                           100   
5                           6                           100   
6                           1                           100   
7                           2                           100   
8                           0                             0   
9                           3                           100   

   COUNT RECEIVES PUBLIC ASSISTANCE  PERCENT RECEIVES PUBLIC ASSISTANCE  \
0                                20                                0.45   
1                                 2                                0.06   
2                                 0                                0.00   
3                                 0                                0.00   
4                                 0                                0.00   
5                                 0                                0.00   
6                                 1                                1.00   
7                                 0                                0.00   
8                                 0                                0.00   
9              

## Level Up (Optional) - Loading Corrupt CSV files

Occasionally, you encountered some really ill formatted data. One example of this can be data that has strings containing commas in a csv file. Under the standard protocol, when this occurs, one is supposed to use quotes to differentiate between the commas denoting fields and commas within those fields themselves. For example, we could have a table like this:  

ReviewerID,Rating,N_reviews,Review,VenueID
123456,4,137,This restaurant was pretty good, we had a great time.,98765

Which should be saved like this if it were a csv (to avoid confusion with the commas in the Review text):
"ReviewerID","Rating","N_reviews","Review","VenueID"
"123456","4","137","This restaurant was pretty good, we had a great time.","98765"

Attempt to import the corrupt file, or at least a small preview of it. It is appropriately titled Yelp_Reviews_corrupt.csv. Investigate some of the intricacies of skipping rows to then pass over this error and comment on what you think is going on.

In [65]:
#Hint: here's a useful programming pattern to use.
try:
    #do something
except Exception as e:
    #handle your exception e

IndentationError: expected an indented block (<ipython-input-65-13f6e15364f1>, line 4)

In [66]:

try:
    df6 = pd.read_csv('Data/Yelp_Reviews_corrupt.csv')
except Exception as e:
    print(e)
    #handle your exception e

Error tokenizing data. C error: Expected 10 fields in line 2331, saw 11



#### The error above saying there are 11 fields when they whould be 10; which means that there might be string that contains a comma that is causing the issue.  

In [67]:
# # Iteration 1 
for i in range(1500,2000):#from the line 1501 to 2000
    try:
        df = pd.read_csv('Data/Yelp_Reviews_Corrupt.csv', nrows=i)#for each iteration, include 1 more additional line
    except:
        print('First failure at: {}'.format(i))#the last line included were causing issue. And the line number is saved in i 
        break
df1 = pd.read_csv('Data/Yelp_Reviews_Corrupt.csv', nrows=i-1)#df1 includes lines before the line where the first error occurred. 
print(len(df))#same as len(df1),the number of lines befor the error
df1.head()

First failure at: 1962
1961


Unnamed: 0             business_id cool        date  funny  \
0          1  pomGBqfbxcqPv14c3XH-ZQ    0  2012-11-13    0.0   
1          2  jtQARsP6P-LbkyjbO1qNGg    1  2014-10-23    1.0   
2          4  Ums3gaP2qM3W1XcA5r6SsQ    0  2014-09-05    0.0   
3          5  vgfcTvK81oD4r50NMjU2Ag    0  2011-02-25    0.0   
4         10  yFumR3CWzpfvTH2FCthvVw    0  2016-06-15    0.0   

                review_id  stars  \
0  dDl8zu1vWPdKGihJrwQbpw    5.0   
1  LZp4UX5zK3e-c5ZGSeo3kA    1.0   
2  jsDu6QEJHbwP2Blom1PLCA    5.0   
3  pfavA0hr3nyqO61oupj-lA    1.0   
4  STiFMww2z31siPY7BWNC2g    5.0   

                                                text  useful  \
0  I love this place! My fiance And I go here atl...     0.0   
1  Terrible. Dry corn bread. Rib tips were all fa...     3.0   
2  Delicious healthy food. The steak is amazing. ...     0.0   
3  This place sucks. The customer service is horr...     2.0   
4  I have been an Emerald Club member for a numbe...     0.0   

                  user_id  
0  msQe1u7Z_XuqjGoqhB0J5g  
1  msQe1u7Z_XuqjGoqhB0J5g  
2  msQe1u7Z_XuqjGoqhB0J5g  
3  msQe1u7Z_XuqjGoqhB0J5g  
4  TlvV-xJhmh7LCwJYXkV-cg

In [68]:
df1.tail()

Unnamed: 0             business_id  \
1956                                         4993  u8C8pRvaHXg3PgDrsUHJHQ   
1957                                         4998  -9nai28tnoylwViuJVrYEQ   
1958    I had an awesome great time with friends.                     NaN   
1959  I loved the tapas and the excellent paella.                     NaN   
1960              I can't wait to come back soon.                       0   

                        cool        date  funny               review_id  \
1956                       0  2016-08-08    0.0  gXmHGBSBBz2-uHdvGf4lZQ   
1957                       0  2015-03-22    0.0  u-zqCN_IXfypJIUzIVUuzw   
1958                     NaN         NaN    NaN                     NaN   
1959                     NaN         NaN    NaN                     NaN   
1960  otDVyX37h61WEbqPLEjCmQ         NaN    NaN                     NaN   

      stars                                               text  useful  \
1956    2.0  just went to a retirement party upstairs and t...     1.0   
1957    5.0             Great restaurant and great atmosphere.     NaN   
1958    NaN                                                NaN     NaN   
1959    NaN                                                NaN     NaN   
1960    NaN                                                NaN     NaN   

                     user_id  
1956  tFa-r1pxZh04FjxNSEQgcQ  
1957                     NaN  
1958                     NaN  
1959                     NaN  
1960                     NaN

###  df1 saved lines before the first error at  the line 1962  

In [72]:
# # Iteration 2 
for i in range(0,500):#for next 500 lines since the error occured
    try:
        temp = pd.read_csv('Data/Yelp_Reviews_Corrupt.csv', skiprows=1962, nrows=i, names=df1.columns)
    except:
        print('First failure at: {}'.format(i))
        break
df2 = pd.read_csv('Data/Yelp_Reviews_Corrupt.csv', skiprows=1962, nrows=i-1, names=df1.columns)
print(len(df2))
df2.head()

498


Unnamed: 0             business_id  \
0  STAY AWAY FROM THIS PLACE!!!!!!                       5   
1                             3948  GAKFx4jFUtTOTpp_jDJnuA   
2                             3949  0QzCeORfF8EY34UODWRV9A   
3                             3950  tlt8zNrZ6_A3DmXiM-cnBA   
4                             3952  XD0LjNuPPwJPsTAHecUh7A   

                     cool        date funny               review_id stars  \
0  sDofYImMQQmu4Le5G9zmpQ         NaN   NaN                     NaN   NaN   
1                       0  2017-09-01     0  OUZWMw7EgO7D596pUelSlA     5   
2                       0  2017-09-03     0  7lbykaWFD8YBwT0mU1Rexw     4   
3                       0  2016-06-12     0  Nd_soHwCYi8adcNIT2w9LQ     1   
4                       0  2015-08-23     0  FUUTAr5CECrkfRa9Y2-MSg     1   

                                                text useful  \
0                                                NaN    NaN   
1  Nice relaxing atmosphere. Friendly service and...      1   
2  Very pleased with our experience. Great off th...      0   
3  Wife went to this location and was horrible. N...      0   
4                        Not baby friendly anymore.     NaN   

                  user_id  
0                     NaN  
1  6vJY67yve43Ijvn8RKVUow  
2  6vJY67yve43Ijvn8RKVUow  
3  S0dnPb1OzaqdBSOxyLr7BQ  
4                     NaN

### df2 saved the lines after the first error (line 1962) and the second (line 1962 + 498)

In [73]:
temp = pd.read_csv('Data/Yelp_Reviews_Corrupt.csv', names=df1.columns, skiprows=1)
print(len(temp))#total number of lines in the file
temp.head()

4651


Unnamed: 0             business_id cool        date funny  \
0          1  pomGBqfbxcqPv14c3XH-ZQ    0  2012-11-13     0   
1          2  jtQARsP6P-LbkyjbO1qNGg    1  2014-10-23     1   
2          4  Ums3gaP2qM3W1XcA5r6SsQ    0  2014-09-05     0   
3          5  vgfcTvK81oD4r50NMjU2Ag    0  2011-02-25     0   
4         10  yFumR3CWzpfvTH2FCthvVw    0  2016-06-15     0   

                review_id stars  \
0  dDl8zu1vWPdKGihJrwQbpw     5   
1  LZp4UX5zK3e-c5ZGSeo3kA     1   
2  jsDu6QEJHbwP2Blom1PLCA     5   
3  pfavA0hr3nyqO61oupj-lA     1   
4  STiFMww2z31siPY7BWNC2g     5   

                                                text useful  \
0  I love this place! My fiance And I go here atl...      0   
1  Terrible. Dry corn bread. Rib tips were all fa...      3   
2  Delicious healthy food. The steak is amazing. ...      0   
3  This place sucks. The customer service is horr...      2   
4  I have been an Emerald Club member for a numbe...      0   

                  user_id  
0  msQe1u7Z_XuqjGoqhB0J5g  
1  msQe1u7Z_XuqjGoqhB0J5g  
2  msQe1u7Z_XuqjGoqhB0J5g  
3  msQe1u7Z_XuqjGoqhB0J5g  
4  TlvV-xJhmh7LCwJYXkV-cg

In [74]:
pd.read_csv('Data/Yelp_Reviews_Corrupt.csv', skiprows=len(df1)+len(df2), names=df1.columns)#the rest of lines don't cause an error

Unnamed: 0  \
0                                                 Cons:   
1     -  Dusty!  Not sure if it's all of Vegas but I...   
2     -  Valet parking: kinda inconvenient when you ...   
3                        -  Sofabed is extremely flimsy   
4                                         Other points:   
5     *  Should call ahead of time to make sure your...   
6                    *  Hotel lobby is extremely small!   
7     *  In-room food service was overpriced (and fo...   
8                               *  Don't go to the 7-11   
9                                               Overall   
10                                                 4058   
11                                                 1624   
12                                                 4938   
13                                                 2897   
14                                  The room was lovely   
15    We took part in the spa package offered throug...   
16    I emailed the concierge ahead of time to reque...   
17    I've given 4 stars instead of 5 for a couple o...   
18                                                 2422   
19                                                 1527   
20                                                 4080   
21                                                 4237   
22                                                 4498   
23                                                  Chi   
24                  Four stars instead of five because:   
25            - The ladies in reception were a bit rude   
26    - My pedicure only lasted two days! A pedicure...   
27                                                 2716   
28                                                 3426   
29    Had the traditional chicken shawarma. one of m...   
...                                                 ...   
2552                           Starting off with drinks   
2553                                 The menu is varied   
2554             We went with the Bambamburger ($11.50)   
2555                  Bamburger serves up great burgers   
2556     If you are hunting for a real deal on a burger   
2557                                               3225   
2558                                               4674   
2559                                               4719   
2560                                               3440   
2561  I had given my husband some ideas of engagemen...   
2562  The customer service here is great - they are ...   
2563  We went back for our wedding bands and I worke...   
2564  We've recommended 5 other friends to H&F - the...   
2565  Go to H&F for all your jewellery needs - you w...   
2566                                               4812   
2567                                               1970   
2568                                                689   
2569                                               4874   
2570                                                564   
2571  Highly recommend the custard cakes they are th...   
2572  The rice flour cake is also really good and a ...   
2573         The bean cakes are great here too!  orange   
2574                       Can't go wrong with Nova Era   
2575                                               3458   
2576                           This was disappointing.    
2577                                          First off   
2578                                             Second   
2579  Finally - Food was mediocre. I was extremely d...   
2580                              Wasn't worth the hype   
2581                                               4206   

                                            business_id  \
0                                                   NaN   
1                                                   NaN   
2                                                   NaN   
3                        if you have more than 2 people   
4                                                   NaN   
5                               

###   After the 2nd error occurring line, the rest of lines were loaded without a problem. 

## Summary

Congratulations, you now practiced your pandas-importing skills!